subworkflows/ebi-metagenomics/fasta_domainannotation/main.nf

```diff
@@ -0,0 +1,45 @@
+include { BLAST_MAKEBLASTDB } from '../../../modules/ebi-metagenomics/blast/makeblastdb/main'
+include { BLAST_BLASTP      } from '../../../modules/ebi-metagenomics/blast/blastp/main'
+include { DIAMOND_MAKEDB    } from '../../../modules/ebi-metagenomics/diamond/makedb/main'
+include { DIAMOND_BLASTP    } from '../../../modules/ebi-metagenomics/diamond/blastp/main'
+include { INTERPROSCAN      } from '../../../modules/ebi-metagenomics/interproscan/main'
+
+workflow FASTA_DOMAINANNOTATION {
+
+    take:
+    ch_fasta       // channel: [ val(meta), path(fasta) ]
+    ch_blast_fasta // channel: /path/to/reference/fasta for blast
+    val_blast_mode // value: blast or diamond
+
+    main:
+
+    ch_versions = Channel.empty()
+
+    if (val_blast_mode == "blast") {
+        println("blast mode")
+        BLAST_MAKEBLASTDB ( ch_blast_fasta )
+        ch_versions = ch_versions.mix(BLAST_MAKEBLASTDB.out.versions)
+        BLAST_BLASTP ( ch_fasta, BLAST_MAKEBLASTDB.out.db, 'tsv' )
+        ch_versions = ch_versions.mix(BLAST_BLASTP.out.versions)
+        blastp_tsv = BLAST_BLASTP.out.tsv
+    } else if (val_blast_mode == "diamond") {
+        println("diamond mode")
+        DIAMOND_MAKEDB ( ch_blast_fasta )
+        ch_versions = ch_versions.mix(DIAMOND_MAKEDB.out.versions)
+        blast_columns = '' // defaults: qseqid, sseqid, pident, length, mismatch, gapopen, qstart, qend, sstart, send, evalue, bitscore
+        DIAMOND_BLASTP ( ch_fasta, DIAMOND_MAKEDB.out.db.map { meta, dmnd -> dmnd }, 'txt', blast_columns )
+        ch_versions = ch_versions.mix(DIAMOND_BLASTP.out.versions)
+        blastp_tsv = DIAMOND_BLASTP.out.txt
+    } else {
+        throw new Exception("Invalid mode value '$val_blast_mode'. Should be 'blast' or 'diamond'.")
+    }
+
+    INTERPROSCAN ( ch_fasta, 'tsv' )
```

Comment:
From our experice, chunking the fasta files for interpro really improves the performance. Interpro with over 4 cores suffers from the io overhead (mostly due to hmmer), so we tend to chunk the input file to use nextflow to paralelize the process.

Reply:
The chunking step happens in the main workflow that calls this subworkflow. Tested and working properly ;)